domain/domain.go

```diff
@@ -1221,6 +1245,128 @@ func (do *Domain) NotifyUpdatePrivilege(ctx sessionctx.Context) {
+
+	logPrefix := fmt.Sprintf("[acquireServerID] ")
+	session, err := owner.NewSession(ctx, logPrefix, do.etcdClient, owner.NewSessionDefaultRetryCnt, cfg.ServerIDTTL)
+	if err != nil {
+		return err
+	}
+	do.serverIDSession.Store(session)
+
+	for i := 0; i < acquireServerIDRetryCnt; i++ {
+		randServerID := rand.Int63n(int64(util.MaxServerID)) + 1 // get a random serverID: [1, MaxServerID]
+		key := fmt.Sprintf("%s/%v", serverIDEtcdPath, randServerID)
+		cmp := clientv3.Compare(clientv3.CreateRevision(key), "=", 0)
+		value := "0"
+
+		childCtx, cancel := context.WithTimeout(ctx, acquireServerIDTimeout)
+		txn := do.etcdClient.Txn(childCtx)
+		t := txn.If(cmp)
+		resp, err := t.Then(clientv3.OpPut(key, value, clientv3.WithLease(do.getServerIDSession().Lease()))).Commit()
+		cancel()
+		if err != nil {
+			return err
+		}
+		if !resp.Succeeded {
+			logutil.BgLogger().Info("random serverID exists, try again", zap.Int64("randServerID", randServerID))
+			time.Sleep(acquireServerIDRetryInterval)
+			continue
+		}
+
+		atomic.StoreUint64(&do.serverID, uint64(randServerID))
+		logutil.BgLogger().Info("acquireServerID", zap.Uint64("serverID", do.ServerID()),
+			zap.String("lease id", strconv.FormatInt(int64(do.getServerIDSession().Lease()), 16)))
+		return nil
+	}
+	logutil.BgLogger().Warn("no serverID available", zap.Int32("retry count", acquireServerIDRetryCnt))
+	return nil
+}
+
+func (do *Domain) refreshServerIDTTL(ctx context.Context) error {
+	key := fmt.Sprintf("%s/%v", serverIDEtcdPath, do.ServerID())
+	value := "0"
+	err := ddlutil.PutKVToEtcd(ctx, do.etcdClient, acquireServerIDRetryCnt, key, value, clientv3.WithLease(do.getServerIDSession().Lease()))
+	if err != nil {
+		logutil.BgLogger().Error("refreshServerIDTTL fail", zap.Uint64("serverID", do.ServerID()), zap.Error(err))
+	} else {
+		logutil.BgLogger().Info("refreshServerIDTTL succeed", zap.Uint64("serverID", do.ServerID()))
+	}
+	return err
+}
+
+func (do *Domain) serverIDKeeper() {
+	cfg := config.GetGlobalConfig()
+	serverIDTimeToCheckPDConnection := time.Duration(cfg.ServerIDTTL) * time.Second / 10
+	lostConnectionTimeout := serverIDTimeToCheckPDConnection * 5
+
+	defer util.Recover(metrics.LabelDomain, "serverIDKeeper", nil, false)
+	ticker := time.NewTicker(serverIDTimeToCheckPDConnection)
+	defer func() {
+		ticker.Stop()
+		do.wg.Done()
+		logutil.BgLogger().Info("serverIDKeeper exited.")
+	}()
+
+	isLostConnectionToPD := false
+	lastSucceedTimestamp := time.Now()
+	for {
+		select {
+		case <-ticker.C:
+			if !isLostConnectionToPD {
+				if err := do.refreshServerIDTTL(context.Background()); err == nil {
+					lastSucceedTimestamp = time.Now()
+				} else {
+					if lostConnectionTimeout > 0 && time.Since(lastSucceedTimestamp) > lostConnectionTimeout {
+						logutil.BgLogger().Warn("lost connection to PD")
+						isLostConnectionToPD = true
+						do.InfoSyncer().GetSessionManager().KillAllConnections()
+					}
+				}
+			}
+		case <-do.getServerIDSession().Done():
```

Comment:
I've not thought about this clearly, shall we `KillAllConnections` or not when this situation happens?